### PR TITLE
Update codegen for fragment implementations

### PR DESCRIPTION
**Pull request by sav007**

### CODE REVIEW

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ast/Ast.kt [hunk 1]
```diff
@@ -209,13 +204,6 @@ internal data class CodeGenerationAst(
   )
 
   companion object {
-    fun customTypeRef(typesPackageName: String): TypeRef {
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ast/Ast.kt:212]: not used anywhere

File: apollo-compiler/src/test/graphql/com/example/arguments_simple/fragment/HeroDetail.kt [hunk 2]
```diff
@@ -73,9 +71,5 @@ interface HeroDetail : GraphqlFragment {
         |  }
         |}
         """.trimMargin()
-
-    operator fun invoke(reader: ResponseReader): HeroDetail {
-      return HeroDetailsImpl_ResponseAdapter.fromResponse(reader)
-    }
```
Comment (martinbonnin) [inline apollo-compiler/src/test/graphql/com/example/arguments_simple/fragment/HeroDetail.kt:79]: 👍

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ir/BackendIr.kt [hunk 3]
```diff
@@ -120,19 +120,19 @@ internal data class BackendIr(
...
-      ) {
+  ) {
     data class SelectionSet(
+        val name: String,
         val fields: List<Field>,
         val fragments: Fragments,
         val typeCondition: IntrospectionSchema.TypeRef,
         val possibleTypes: Set<IntrospectionSchema.TypeRef>,
+        val defaultSelectionKey: SelectionKey,
         val selectionKeys: Set<SelectionKey>,
```
Comment (martinbonnin) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/backend/ir/BackendIr.kt:136]: Does `selectionKeys` need to be a `Set`? I only see usages of it containing one value. It actually looks like it's always equal to `defaultSelectionKey` ? Or do I miss something?

**Comment by martinbonnin** [comment]
Looks nice 👍 ! I took the liberty of pushing a couple of fixes for tests.

More generally, should we put the fragments implementation generation behind an `opt-in` flag? My understanding is that this will only be used to read/write fragments to the store, which is an advanced feature. We might save a lot of users some confusion and build time if we disable it by default?

**Pull request merged**